backend/msgpack/conn.go

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"errors"
+	"math"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3x/clock"
+)
+
+const (
+	tcpProtocol = "tcp"
+)
+
+var (
+	errNoActiveConnection = errors.New("no active connection")
+)
+
+type connectWithLockFn func() error
+type writeWithLockFn func([]byte) error
+
+// connection is a persistent connection that retries establishing
+// connection with exponential backoff if the connection goes down.
+type connection struct {
+	sync.Mutex
+
+	addr         string
+	opts         ConnectionOptions
+	connTimeout  time.Duration
```

Comment:
I guess my point was that why not either:

1) Pull everything you need from the options to their own fields
2) Leave everything you need inside opts.

Seems odd to pull some properties out but leave others to be accessed inside opts.